examples/async-data/src/App.js

```diff
@@ -0,0 +1,186 @@
+  delay: 10,
+});
+
+class App extends Component {
+  constructor() {
+    super();
+
+    // A basic starting plotly.js figure object. Instead of assigning
+    const figure = {
+      data: [{type: 'scatter'}],
+      layout: {title: 'Room readings'},
+    };
+
+    // Store the figure, dataSource and dataSourceOptions in state.
+    this.state = {figure, dataSources: {}};
+
+    // Instantiate a Hub object. The hub is a convenience module that updates
+    // the applies Editor updates to the figure object. After an update it
+    // will call the onUpdate function with the editor and plot revision numbers.
+    // We set these in our state and pass them to react-plotly.js-editor and
+    // react-plotly.js plot component respectively. This is necessary because
+    // the plotly.js library will mutate the figure object with user interactions.
+    // The hub listens for events from plotly.js and alerts us to the mutation here.
+    // The Editor follows the same mutation pattern (for good or ill) and the Hub
+    // will pick up editor results in the same way.
+    this.hub = new Hub({
+      debug: true,
+      onUpdate: ({editorRevision, plotRevision}) =>
+        this.setState(prevState => ({
+          ...prevState,
+          editorRevision,
+          plotRevision,
+        })),
+    });
+
+    this.getChartingData = this.getChartingData.bind(this);
+    this.setChartingData = this.setChartingData.bind(this);
+    this.setChartingDataOptions = this.setChartingDataOptions.bind(this);
+    this.onEditorUpdate = this.onEditorUpdate.bind(this);
+  }
+
+  componentDidMount() {
+    backend.on('ChartingDataColumnNames', this.setChartingDataOptions);
+    backend.on('ChartingData', this.setChartingData);
+
+    this.getChartingDataColumnsNames();
+  }
+
+  componentDidUnmount() {
+    backend.off('ChartingDataColumnNames', this.setChartingDataOptions);
+    backend.off('ChartingData', this.setChartingData);
+  }
+
+  setChartingDataOptions(columnNames) {
+    const dataSourceOptions = columnNames.map(name => ({
+      value: name,
+      label: name,
+    }));
+    this.setState(prevState => ({...prevState, dataSourceOptions}));
+    this.hub.editorSourcesUpdated();
+  }
+
+  getChartingDataColumnsNames() {
+    backend.fetch({
+      resource: 'GetChartingDataColumnNames',
+    });
+  }
+
+  getChartingData(columnName) {
+    backend.fetch({
+      resource: 'GetChartingData',
+      data: {columnName},
+    });
+  }
+
+  setChartingData({columnName, data}) {
+    if (Array.isArray(data) && data.length) {
+      const {dataSources, ...otherState} = this.state;
+      dataSources[columnName] = data;
+      dereference(this.state.figure.data, dataSources);
+      this.setState({
+        dataSources,
+        ...otherState,
+      });
+      this.hub.figureUpdated();
+    }
+  }
+
+  onEditorUpdate(event) {
+    const {type, payload} = event;
+    if (type === EDITOR_ACTIONS.UPDATE_TRACES) {
```

Comment:
And if we need to intercept the figure updating routine you gotta know how EDITOR_ACTIONS works and which events to tie into ... please :see_no_evil:

Reply:
here we could invert control and have the hub fire its own sub-events that apps could hook into, so this would be implemented as `hub.onUpdateTrace = ...`

Reply:
can we eliminate the hub entirely? It still seems a bit ungainly to me. It's a stateful thing but distinct from typical application data handling.